content/blog/performant-elm/index.md

```diff
@@ -0,0 +1,135 @@
+more information about what it means
+[here](https://developers.google.com/web/fundamentals/performance/rail#goals-and-guidelines "Measure Performance with the RAIL Model").
+
+You'll also notice you can select a smaller time interval and start digging
+into more specific bits of what your application is doing at a certain
+time. For example, if you highlight the first big chunk of scripting and
+then look in the lower panel, you will find a tab named "Bottom-Up".
+
+![Bottom-up](./bottom-up.png)
+
+The first entry says `_Utils_eqHelp`, and apparently that function took the
+most time to execute. This is curious, because we don't expect our
+application to spend so much time comparing things. By clicking on the
+right hand side, we will be able to see which module is the cause of this
+performance problem.
+
+![Source](./source.png)
+
+As you can see, the function causing slowdown seems to be the aptly named
+`author$project$Page$Learn$SelfReviews$Revise$Highlighter$isSpace`. Let's
+look at what that function does:
+
+```elm
+isSpace : Char -> Bool
+isSpace char =
+    List.member char
+        [ ' '
+        , '\t'
+        , '\n'
+        , '\u{000D}'
+
+        -- what follows is a list of unicode characters used
+        -- as spaces: http://jkorpela.fi/chars/spaces.html
+        , '\u{00A0}'
+        , '\u{1680}'
+        , '\u{180E}'
+        , '\u{2000}'
+        , '\u{2001}'
+        , '\u{2002}'
+        , '\u{2003}'
+        , '\u{2004}'
+        , '\u{2005}'
+        , '\u{2006}'
+        , '\u{2007}'
+        , '\u{2008}'
+        , '\u{2009}'
+        , '\u{200A}'
+        , '\u{200B}'
+        , '\u{202F}'
+        , '\u{205F}'
+        , '\u{3000}'
+        , '\u{FEFF}'
+        ]
+```
+
+Uh! That is really surprising. Usually iterating over a list is really
+fast, but in this case it seems that we are spending almost 90% of our
+time inside this function. As you probably know, the algorithmic complexity
+of iterating over a list is _O(n)_: we could easily improve that by
+building a set that contains all the desired characters and checking
+if the current character is part of that set. The Elm standard library
+implements a set using a dict, which is in turn implemented using a
+red-black tree, which means that the time complexity will become
+_O(log(n))_.
+
+So we can change the code that looked like this
+
+```elm
+sameToken ( _, char1 ) ( _, char2 ) =
+    not (isSpace char1) && not (isSpace char2)
+```
+
+into something that looks like this
+
+```elm
+-- This is a very hot path in the code. So we build a set once
+-- and use it to determine if a character is a space or not.
+
+spaceSet =
+    Set.fromList spaceCharacters
+
+sameToken ( _, char1 ) ( _, char2 ) =
+    not (Set.member char1 spaceSet) &&
+      not (Set.member char2 spaceSet)
+```
+
+With this little change in place, we can repeat the performance
+measurement
+
+![Append Child](./append-child.png)
```

Comment:
Consider a way of showing a delta between this screenshot and the previous screenshot while drawing attention to the performance improvement. Is the reader supposed to notice fewer red corners? An improved Bottom-Up column? A couple of ideas: 1. A split view before / after with the 'after' screenshot annotated with where the user should be looking. 2. A gif that toggles between before / after in a loop.